File: dynamic_listings/python/default/containers/fsm.py
from typing import Callable


class FSM:
    """
    This class defines a Finite State Machine
    The currently active state is represented by a function
    pointer
    """

    current_state: Callable = None

    def set_state(self, f: Callable) -> None:
        """
        Sets the state, from this point on, update will
        change its strategy
        """
        self.current_state = f

    def update(self, dt: float) -> None:
        # If there is a current state, execute it
        if self.current_state:
            self.current_state(dt)

File: dynamic_listings/C++/default/containers/fsm.cpp
#include <functional>

class FSM{
    /*
     * This class defines a Finite State Machine
     * The currently active state is represented by a function
     * pointer
     */

    std::function<void(float)> current_state = nullptr;

    void setState(std::function<void(float)> f){
        /*
         * Sets the state, from this point on, update will
         * change its strategy
         */
        current_state = f;
    }

    void update(float dt){
        // If there is a current state, execute it
        if (current_state != nullptr){
            current_state(dt);
        }
    }
};
